Adds offices tests

anselmbradford opened the pull request:
Adds offices tests

## Additions

- Adds page object and shared spec suite for /offices/ pages content.

## Testing

- Follow `TEST.md` instructions for browser tests.

## Review

- @imuchnik 
- @jimmynotjim 
- @KimberlyMunoz 
- @sebworks 

## Notes

- ~~So when I run these I get `FATAL ERROR: CALL_AND_RETRY_2 Allocation failed - process out of memory` and the Selenium browser instances remain open. A few notes:~~

 - ~~The selectors are very long because I leaned toward not cluttering the markup with classes that are just used for test suite hooks. It didn't seem problematic to have them long since they can be recreated with inspect element > Copy CSS Path. But could this be the issue?~~ Fixed with `qa-` selectors.

 - There are three different page options in the get method of the page object because not all office pages have all the possible content available.

imuchnik commented on test/browser_tests/spec_suites/shared/shared_office.js:
```diff
@@ -0,0 +1,22 @@
+describe( 'The offices template', function() {
+  it( 'should properly load in a browser', function() {
+    var page = new Office();
+    page.get( 'ProjectCatalyst' );
```
This line is duplicated for every test. Consider moving into the Page Object method and invoke it in beforeEach function in the spec

anselmbradford replied:
So the reason it is here instead of `beforeEach` is some tests will use `page.get( 'PlainWriting' )` and `page.get( 'OfficeOfFinancialEducation' )`. I had all three in `beforeEach` and that meant that every test was instantiating two more Office objects than it needed. I'm not sure if it would make sense to maybe have three `describe` blocks for each type of office page, and the tests associated with each. All three are needed because that's the only way I could get full coverage of the office page content.

jimmynotjim replied:
I wonder if it's worth breaking up in the page objects like I did, but instead of repeating the ones that are shared across pages, there's a function that get's passed the parent element (like this.eventElem in my PR) and returns an object with the shared elems saved.

KimberlyMunoz commented:
I think adding `.qa-` classes or at least `.office-` specific classes would be a better way to go even if it isn't the issue because those selectors are very fragile if anything changes on the page. 

I'll pull this down to look at today to try a few things.

jimmynotjim commented:
I found yesterday that the tests hang if it can't find a selector. I was purposefully looking for the map on the event-archive, knowing it shouldn't exist on that page, and every time it would hang. You'll have to test stuff that you know will exist for the time being till I can find a solution (currently doing that).

imuchnik commented:
Yes, XPath selectors are probably the slowest and most fragile. Usually selecting by id is the fastest way followed by css selecotrs

KimberlyMunoz commented:
I played with this a little. Changing to smaller selectors avoids the "out of memory" problem and will probably lead to happier shorter tests down the road.

anselmbradford commented:
> I played with this a little. Changing to smaller selectors avoids the "out of memory" problem and will probably lead to happier shorter tests down the road.

Cool. Okay, sounds like that's the issue. Is there any other way than adding classes for the tests? I'm not very into the idea of classes that are in the project solely for the purpose of test hooks :(

KimberlyMunoz commented:
I don't think so. :cry: 

I think having classes for tests is fine, but [I just would want to label them `.qa-` so that we know they're just for testing and they can be kept pure or refactored into existing classes later.](http://csswizardry.com/2015/03/more-transparent-ui-code-with-namespaces/) It seems like we already lean towards more classes to make our CSS cleaner. I think added classes to make our tests quicker and more robust is in line with that.

jimmynotjim commented:
:point_up:

anselmbradford commented:
Pushed edits in https://github.com/cfpb/cfgov-refresh/commit/9574c57bd3d3c1137dd7185fb86d00396a1b0783 but it still crashes for me:

<img width="804" alt="screen shot 2015-07-27 at 10 51 39 am" src="https://cloud.githubusercontent.com/assets/704760/8909202/2b4dddfe-344e-11e5-8c3c-ba7fad8d541f.png">

After a rebase with flapjack:
<img width="790" alt="screen shot 2015-07-27 at 10 55 49 am" src="https://cloud.githubusercontent.com/assets/704760/8909206/30c30cf0-344e-11e5-8e6c-ddff0d6438ea.png">

jimmynotjim commented:
@anselmbradford did you remove the page objects that don't exist across the pages you're testing?

anselmbradford commented:
> did you remove the page objects that don't exist across the pages you're testing?

Only the ProjectCatalyst page is being accessed for the tests thus far, is that what you mean? Haven't looked at how you set this up in the events yet per https://github.com/cfpb/cfgov-refresh/pull/786#discussion_r35131949

anselmbradford commented:
This one is ready for final review and sign-off now. I added the office pages we have that are accessible through the site's menus and removed the office of financial education, since that isn't an officially released page yet.

sebworks commented:
Is there a reason why we are using `toBeTruthy/toBeFalsy` instead of `toBe(true)/toBe(false)`?

https://github.com/angular/protractor/blob/2.1.0/lib/element.js#L923

anselmbradford commented:
Oops, pushed a slight fix

anselmbradford commented:
> Is there a reason why we are using toBeTruthy/toBeFalsy instead of toBe(true)/toBe(false)?

Not sure, I was going with existing conventions I saw. What's the difference?

sebworks commented:
`!!` will coerce the value to be Boolean. Using `truthy/falsy` suggests that an alternative value is possible.

anselmbradford commented:
> Is there a reason why we are using toBeTruthy/toBeFalsy instead of toBe(true)/toBe(false)?

Fixed in https://github.com/cfpb/cfgov-refresh/commit/ff91be16d4a87c5b0a29fcd18feb1254d72b4c69

sebworks commented:
:+1:
